### PR TITLE
fix(gauge): Remove current_index related to ring buffer vectors

### PR DESCRIPTION
**Pull request by quangkeu95**
Struct `Gauge`, `GaugeVoter` and `GaugeVote` contain a field which type is a vector that implements the ring buffer pattern. We want to insert an item into the vector at the index that relative with the voting epoch number. For example, voting epoch 1 at index 0, voting epoch 2 at index 1, no voting epoch at index 2 and voting epoch 3 at index 3, so on and so forth. 

Tasks:
- [x] Update formula to calculate index for the voting epoch.
- [x] Update tests.

### CODE REVIEW

File: programs/gauge/src/instructions/pump_gauge_epoch.rs [hunk 1]
```diff
@@ -62,10 +62,10 @@ pub fn handler(ctx: Context<PumpGaugeEpoch>) -> Result<()> {
     let index = gauge.pump_and_get_index_for_lastest_voting_epoch(voting_epoch);
     let vote_epoch = &mut gauge.vote_epochs[index];
 
-    require!(
```
Comment (andrewsource147) [inline programs/gauge/src/instructions/pump_gauge_epoch.rs:65]: why comment here?

File: programs/gauge/src/state.rs [hunk 2]
```diff
@@ -174,14 +173,14 @@ impl GaugeVote {
...
+        let index = (latest_voting_epoch % max_epoch_per_gauge) as usize;
+
+        if self.vote_epochs[index].voting_epoch == latest_voting_epoch {
+            return index;
         }
-        let current_index = current_index.checked_add(1).unwrap() % MAX_EPOCH_PER_GAUGE;
-        self.current_index = u64::try_from(current_index).unwrap();
-        self.vote_epochs[current_index].voting_epoch = latest_voting_epoch;
-        self.current_index.try_into().unwrap()
+        self.vote_epochs[index].voting_epoch = latest_voting_epoch;
```
Comment (andrewsource147) [inline programs/gauge/src/state.rs:182]: do we need to reset other fields?

File: programs/gauge/src/state.rs [hunk 3]
```diff
@@ -332,13 +323,14 @@ impl Gauge {
...
+        let max_epoch_per_gauge: u32 = u32::try_from(MAX_EPOCH_PER_GAUGE).unwrap();
+        let index = (latest_voting_epoch % max_epoch_per_gauge) as usize;
+
+        if self.vote_epochs[index].voting_epoch == latest_voting_epoch {
+            return index;
         }
-        let current_index = current_index.checked_add(1).unwrap() % MAX_EPOCH_PER_GAUGE;
-        self.current_index = u64::try_from(current_index).unwrap();
-        self.current_index.try_into().unwrap()
+        self.vote_epochs[index].voting_epoch = latest_voting_epoch;
```
Comment (andrewsource147) [inline programs/gauge/src/state.rs:332]: ditto^

File: programs/gauge/src/state.rs [hunk 4]
```diff
@@ -395,19 +387,22 @@ impl GaugeVoter {
         return Err(VotingEpochNotFound.into());
     }
 
-    // should return if it not pump
+    /// voting epoch start from 1
+    /// max index that we can store voting epoch is MAX_EPOCH_PER_GAUGE
+    /// so index_for_latest_voting_epoch = latest_voting_epoch % MAX_EPOCH_PER_GAUGE - 1 if latest_voting_epoch != MAX_EPOCH_PER_GAUGE
+    /// else index_for_latest_voting_epoch = MAX_EPOCH_PER_GAUGE - 1
```
Comment (andrewsource147) [inline programs/gauge/src/state.rs:393]: we should remove comment here? It is incorrect

File: programs/gauge/src/instructions/pump_gauge_epoch.rs (hunk 1 above)
Comment (quangkeu95) [reply programs/gauge/src/instructions/pump_gauge_epoch.rs:53]: updated

File: programs/gauge/src/state.rs (hunk 4 above)
Comment (quangkeu95) [reply programs/gauge/src/state.rs]: removed

File: programs/gauge/src/state.rs (hunk 2 above)
Comment (quangkeu95) [reply programs/gauge/src/state.rs]: changed: made functions that doesn't modify the `vote_epochs` to take a `&self`

File: programs/gauge/src/state.rs (hunk 3 above)
Comment (quangkeu95) [reply programs/gauge/src/state.rs]: changed: made functions that doesn't modify the `vote_epochs` to take a `&self`

File: programs/gauge/src/state.rs (hunk 2 above)
Comment (quangkeu95) [reply programs/gauge/src/state.rs]: functions that modify the `vote_epochs` should reset the inner item at index

File: programs/gauge/src/state.rs [hunk 5]
```diff
@@ -503,9 +485,16 @@ impl EpochBribeVoter {
         gauge_voter: &GaugeVote,
         gauge: &Gauge,
     ) -> Result<u64> {
+        msg!(
```
Comment (andrewsource147) [inline programs/gauge/src/state.rs:488]: pls remove log here

**Pull request merged**